libsupport/include/katana/Result.h

```diff
@@ -394,6 +400,31 @@ operator<<(std::ostream& out, const CopyableErrorInfo& ei) {
   return ei.Write(out);
 }
 
+}  // namespace katana
+
+// We inject the trait as per https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/reference/traits/is_error_type.html
+// "Overridable: By template specialisation into the trait namespace."
+// It is safer to define the trait before the definition of the Result structure.
+BOOST_OUTCOME_V2_NAMESPACE_BEGIN
+
+namespace trait {
+
+template <>
+struct is_error_type<katana::ErrorInfo> {
+  static constexpr bool value = true;
```

Comment:
The whole fix lets the GPU build succeed, see https://github.com/KatanaGraph/katana-enterprise/pull/1877 (the latest GPU jobs are still in the queue, but a previous run built the sources ok).
This specific place doesn't fix all the `nvcc` issues and further fixes are necessary, however, it is the right thing to do because earlier we were using Boost interface incorrectly: without setting a trait it needs. With the trait in place, it relaxes some requirements, see "Protected member predicate booleans" at https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/reference/types/basic_result.html

Reply:
```
predicate::implicit_constructors_enabled is constexpr boolean true if:

    1. predicate::constructors_enabled is true.
    2. Trait is_error_type<E> is not true for both decayed value_type and decayed error_type at the same time.
    3. value_type is not implicitly constructible from error_type and error_type is not implicitly constructible from value_type.
    OR
    trait is_error_type<E> is true for decayed error_type and error_type is not implicitly constructible from value_type and value_type is an integral type.
```

Reading the logic, it sounds like `is_error_type` only becomes relevant if `error_type` is an integral type. As for common non-integral types, we will fall into case 3a: `value_type` is not constructible from `error_type` and `error_type` is not constructable from `value_type`.

Anyway, I don't have a problem with adding the trait but I doubt it does anything for us.

While you are here can you apply the changes?

```
diff --git a/libsupport/include/katana/EntityTypeManager.h b/libsupport/include/katana/EntityTypeManager.h
index 22b17d65d..bd30cec9f 100644
--- a/libsupport/include/katana/EntityTypeManager.h
+++ b/libsupport/include/katana/EntityTypeManager.h
@@ -232,24 +232,7 @@ public:
   Result<EntityTypeID> GetOrAddEntityTypeID(const std::string& name);
 
   Result<katana::TypeNameSet> EntityTypeToTypeNameSet(
-      katana::EntityTypeID type_id) const {
-    katana::TypeNameSet type_name_set;
-    if (type_id == katana::kInvalidEntityType ||
-        type_id >= GetNumEntityTypes()) {
-      return KATANA_ERROR(
-          ErrorCode::InvalidArgument,
-          "no string representation for invalid type");
-    }
-    auto type_set = GetAtomicSubtypes(type_id);
-    for (size_t idx = 0; idx < type_set.size(); ++idx) {
-      if (type_set[idx]) {
-        auto name = GetAtomicTypeName(idx);
-        KATANA_LOG_ASSERT(name.has_value());
-        type_name_set.insert(name.value());
-      }
-    }
-    return type_name_set;
-  }
+      katana::EntityTypeID type_id) const;
 
   /// \returns the EntityTypeIDs for atomic types with \p names, or an error if
   /// any does not exist.
@@ -269,7 +252,7 @@ public:
             ErrorCode::NotFound, "type {} does not exist", name);
       }
     }
-    return res;
+    return MakeResult(std::move(res));
   }
 
   /// \returns the EntityTypeIDs for atomic types with \p names, adding them if
@@ -290,7 +273,7 @@ public:
       }
       res[id] = true;
     }
-    return res;
+    return MakeResult(std::move(res));
   }
 
   /// \returns the name of the atomic type if the EntityTypeID
diff --git a/libsupport/include/katana/Result.h b/libsupport/include/katana/Result.h
index a9cb0efa1..e5b1a2cf6 100644
--- a/libsupport/include/katana/Result.h
+++ b/libsupport/include/katana/Result.h
@@ -9,6 +9,7 @@
 
 #include <arrow/result.h>
 #include <boost/outcome/outcome.hpp>
+#include <boost/outcome/trait.hpp>
 #include <boost/outcome/utils.hpp>
 #include <fmt/format.h>
 
@@ -225,10 +226,10 @@ public:
 
   static constexpr int kContextSize = 512;
 
-  ErrorInfo(const std::error_code& ec) : error_code_(ec) {}
-
   ErrorInfo() : ErrorInfo(std::error_code()) {}
 
+  ErrorInfo(const std::error_code& ec) : error_code_(ec) {}
+
   template <
       typename ErrorEnum, typename U = std::enable_if_t<
                               std::is_error_code_enum_v<ErrorEnum> ||
@@ -343,6 +344,13 @@ public:
 
   CopyableErrorInfo(const ErrorInfo& ei);
 
+  template <
+      typename ErrorEnum, typename U = std::enable_if_t<
+                              std::is_error_code_enum_v<ErrorEnum> ||
+                              std::is_error_condition_enum_v<ErrorEnum>>>
+  CopyableErrorInfo(ErrorEnum&& err)
+      : CopyableErrorInfo(make_error_code(std::forward<ErrorEnum>(err))) {}
+
   template <typename F, typename... Args>
   CopyableErrorInfo WithContext(F&& fmt_string, Args&&... args) {
     PrependFmt(std::forward<F>(fmt_string), std::forward<Args>(args)...);
@@ -394,6 +402,30 @@ operator<<(std::ostream& out, const CopyableErrorInfo& ei) {
   return ei.Write(out);
 }
 
+}  // namespace katana
+
+// Tell boost::outcome which types will be used as an error type E in
+// std_result<T, E, ...> below.
+BOOST_OUTCOME_V2_NAMESPACE_BEGIN
+
+namespace trait {
+
+template <>
+struct is_error_type<katana::ErrorInfo> {
+  static constexpr bool value = true;
+};
+
+template <>
+struct is_error_type<katana::CopyableErrorInfo> {
+  static constexpr bool value = true;
+};
+
+}  // namespace trait
+
+BOOST_OUTCOME_V2_NAMESPACE_END
+
+namespace katana {
+
 /// make_error_code converts ErrorInfo into a standard error code. It is an STL
 /// and boost::outcome extension point and will be found with ADL if necessary.
 inline std::error_code
@@ -452,7 +484,7 @@ operator!=(const CopyableErrorInfo& a, const CopyableErrorInfo& b) {
 KATANA_EXPORT Result<void> ResultSuccess();
 KATANA_EXPORT CopyableResult<void> CopyableResultSuccess();
 // TODO(ddn): nvcc has trouble applying implicit conversions while other
-// compilers can apply the implicit conversion from ErrorInfo to Result.R emove
+// compilers can apply the implicit conversion from ErrorInfo to Result. Remove
 // this function when nvcc improves.
 KATANA_EXPORT Result<void> ResultError(ErrorInfo&& info);
 
```